beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,255 @@
+- [Release Plan](#release-plan)
+- [Dependencies](#dependencies)
+- [Alternatives](#alternatives)
+
+## Summary
+
+Add a core `MetricsService` to Backstage's framework that provides a unified interface for metrics instrumentation. The service offers OpenTelemetry-based capabilities with support for app configuration (e.g. `app-config.yaml`).
+
+## Motivation
+
+There is currently no guidance for plugin authors when it comes to metrics instrumentation. While individual plugins may implement their own metrics, there's no standardized approach for metrics collection, naming, or configuration.
+
+By providing a core metrics service:
+
+- **Plugin Authors** and the **Community** gain a straightforward way to address metrics instrumentation and can focus on business logic instead of needing to reimplement metrics plumbing.
+- **Backstage Admins** receive a reliable stream of metrics from the core system for monitoring, alerting, and troubleshooting.
+
+### Goals
+
+- Consistent metrics patterns across all plugins
+- Aligned with OpenTelemetry industry standards
+- Provide a familiar interface as other core services
+
+The catalog and scaffolder plugins will be updated to use the new metrics service in the initial release.
+
+### Non-Goals
+
+- Adding metrics to plugins that don't currently have it (outside of catalog and scaffolder)
+- Providing a full Backstage telemetry SDK (aka recreate the OTEL Node SDK).
+- Tracing and other telemetry concerns are out of scope for this BEP.
+- Refactoring the existing `LoggerService`. Future work to unify observability related concerns would be ideal, but not a goal.
+
+## Proposal
+
+Following similar patterns to other core services, create a new `RootMetricsService` responsible for initializing metrics, root-level concerns, and the creation of plugin-specific `MetricsService` instances. The root service delegates to the plugin-scoped `MetricsService` to initialize a meter for each registered plugin based on the `serviceName` and optional `serviceVersion` provided in the app's `app-config.yaml` file. This is based on the recommendation in the OpenTelemetry [documentation](https://opentelemetry.io/docs/languages/js/instrumentation/#acquiring-a-meter).
+
+## Design Details
+
+### Configuration
+
+<!--
+todo: I want to provide full configuration via the app-config, but that requires us wrapping the entire SDK and handling the `instrumentation.js`
+      instructions. I'm not sure we want to start there.
+
+Metrics configuration will continue to be defined by the adopter in their `instrumentation.js` file as outlined in the [setup-opentelemetry](https://backstage.io/docs/tutorials/setup-opentelemetry/) docs.
+
+The initial alpha release will include a simple configuration for the root metrics service.
+-->
+
+```yaml
+backend:
+  metrics:
+    serviceName: backstage
+    # todo: this is optional - should we just default it to the current Backstage release and allow override?
+    serviceVersion: 0.0.1
```

Comment:
Yeah, the Backstage version seems a reasonable default but Adopters may also tag their releases and that version would be handy here.

Reply:
Marking resolved since configuration has been marked out of scope.